crates/wasmtime/src/runtime/store.rs

```diff
@@ -1518,8 +1584,102 @@ impl StoreOpaque {
+            // Aim to double the heap size, amortizing the cost of growth.
+            let doubled_size_in_pages = current_size_in_pages.saturating_mul(2);
+            assert!(doubled_size_in_pages >= current_size_in_pages);
+            let delta_pages_for_doubling = doubled_size_in_pages - current_size_in_pages;
+
+            // When doubling our size, saturate at the maximum memory size in pages.
+            //
+            // TODO: we should consult the instance allocator for its configured
+            // maximum memory size, if any, rather than assuming the index
+            // type's maximum size.
+            let max_size_in_bytes = 1 << 32;
+            let max_size_in_pages = max_size_in_bytes / page_size;
+            let delta_to_max_size_in_pages = max_size_in_pages - current_size_in_pages;
+            let delta_pages_for_alloc = delta_pages_for_doubling.min(delta_to_max_size_in_pages);
+
+            // But always make sure we are attempting to grow at least as many pages
+            // as needed by the requested allocation. This must happen *after* the
+            // max-size saturation, so that if we are at the max already, we do not
+            // succeed in growing by zero delta pages, and then return successfully
+            // to our caller, who would be assuming that there is now capacity for
+            // their allocation.
+            let pages_needed = bytes_needed.div_ceil(page_size);
+            assert!(pages_needed > 0);
+            let delta_pages_for_alloc = delta_pages_for_alloc.max(pages_needed);
+            assert!(delta_pages_for_alloc > 0);
+
+            // Safety: we pair growing the GC heap with updating its associated
+            // `VMMemoryDefinition` in the `VMStoreContext` immediately
+            // afterwards.
+            unsafe {
+                memory
+                    .grow(delta_pages_for_alloc, Some(self.traitobj().as_mut()))?
+                    .ok_or_else(|| anyhow!("failed to grow GC heap"))?;
+            }
+            self.vm_store_context.gc_heap = memory.vmmemory();
+
+            let new_size_in_bytes = u64::try_from(memory.byte_size()).unwrap();
+            assert!(new_size_in_bytes > current_size_in_bytes);
+            delta_bytes_grown = new_size_in_bytes - current_size_in_bytes;
+            let delta_bytes_for_alloc = delta_pages_for_alloc.checked_mul(page_size).unwrap();
+            assert!(
+                delta_bytes_grown >= delta_bytes_for_alloc,
+                "{delta_bytes_grown} should be greater than or equal to {delta_bytes_for_alloc}"
+            );
+            Ok(())
+        })();
+
+        // Regardless whether growing succeeded or failed, place the memory back
+        // inside the GC heap.
+        self.unwrap_gc_store_mut()
+            .gc_heap
+            .replace_memory(memory, delta_bytes_grown);
+
+        grow_result
+    }
+
+    #[cfg(feature = "gc")]
+    pub fn grow_or_collect_gc_heap(&mut self, bytes_needed: Option<u64>) {
```

Comment:
re my comment below: this was my worry, the method here duplicates `maybe_async_grow_or_collect_gc_heap` and I find it confusing to trace what method is used where. I'd prefer ideally if there were a single "do gc" method that everything funnels through rather than having a forest of methods with no narrow waist